core/network-libp2p/src/service_task.rs

```diff
@@ -221,10 +221,71 @@ struct NodeInfo {
 	endpoint: ConnectedPoint,
 	/// Version reported by the remote, or `None` if unknown.
 	client_version: Option<String>,
+	/// Latest ping time with this node.
+	latest_ping: Option<Duration>,
 }
 
 impl<TMessage> Service<TMessage>
 where TMessage: CustomMessage + Send + 'static {
+	/// Returns a struct containing tons of useful information about the network.
+	pub fn state(&mut self) -> NetworkState {
+		let now = Instant::now();
+
+		let connected_peers = {
+			let swarm = &mut self.swarm;
+			self.nodes_info.values().map(move |info| {
+				let known_addresses = swarm.known_addresses(&info.peer_id)
+					.map(|(a, s)| (a.clone(), s)).collect();
+
+				(info.peer_id.to_base58(), NetworkStatePeer {
+					endpoint: match info.endpoint {
```

Comment:
Maybe move this match code into a `Into<NetworkStatePeerEndpoint> for ConnectedPoint`?